src/main/java/net/dv8tion/jda/api/entities/emoji/Emoji.java

```diff
@@ -0,0 +1,236 @@
+     *
+     * @throws IllegalArgumentException
+     *         If the name is null or empty
+     *
+     * @return The new emoji instance
+     */
+    @Nonnull
+    static CustomEmoji fromCustom(@Nonnull String name, long id, boolean animated)
+    {
+        Checks.notEmpty(name, "Name");
+        return new EmojiImpl(name, id, animated);
+    }
+
+    /**
+     * Creates a reference for a custom emoji from the provided {@link CustomEmoji}
+     *
+     * @param  emoji
+     *         The emoji instance
+     *
+     * @throws IllegalArgumentException
+     *         If the emoji is null
+     *
+     * @return The new emoji instance
+     */
+    @Nonnull
+    static CustomEmoji fromCustom(@Nonnull CustomEmoji emoji)
+    {
+        Checks.notNull(emoji, "Emoji");
+        return fromCustom(emoji.getName(), emoji.getIdLong(), emoji.isAnimated());
+    }
+
+    // either <a?:name:id> or just unicode
+
+    /**
+     * Parses the provided markdown formatting, or unicode characters, to an Emoji instance.
+     * <h4>Example</h4>
+     * <pre>{@code
+     * // animated custom emoji
+     * fromFormatted("<a:dance:123456789123456789>");
+     * // not animated custom emoji
+     * fromFormatted("<:dog:123456789123456789>");
+     * // unicode emoji, escape codes
+     * fromFormatted("&#92;uD83D&#92;uDE03");
+     * // codepoint notation
+     * fromFormatted("U+1F602");
+     * // unicode emoji
+     * fromFormatted("😃");
+     * }</pre>
+     *
+     * @param  code
+     *         The code to parse
+     *
+     * @throws IllegalArgumentException
+     *         If the provided code is null or empty
+     *
+     * @return The emoji instance
+     */
+    @Nonnull
+    static Emoji fromFormatted(@Nonnull String code)
+    {
+        Matcher matcher = Message.MentionType.EMOJI.getPattern().matcher(code);
+        if (matcher.matches())
+            return fromCustom(matcher.group(1), Long.parseUnsignedLong(matcher.group(2)), code.startsWith("<a"));
+        else
+            return fromUnicode(code);
+    }
+
+    /**
+     * Parses the provided JSON representation to an emoji instance.
+     *
+     * @param  emoji
+     *         The emoji json
+     * @throws net.dv8tion.jda.api.exceptions.ParsingException
+     *         If the JSON is not a valid emoji
+     *
+     * @return The emoji instance
+     */
+    @Nonnull
+    static Emoji fromData(@Nonnull DataObject emoji)
+    {
+        return new EmojiImpl(emoji.getString("name"),
+                emoji.getUnsignedLong("id", 0),
+                emoji.getBoolean("animated"));
+    }
+
+    /**
+     * The {@link Type} of this emoji.
+     *
+     * @return The {@link Type}
+     */
+    @Nonnull
+    Type getType();
+
+    /**
+     * The name of this emoji.
+     * <br>This will be the unicode characters if this emoji is not of {@link #getType()} {@link Type#CUSTOM CUSTOM}.
+     *
+     * @return The unicode or custom name
+     */
+    @Nonnull
+    String getName();
```

Comment:
Given the existence of `getFormatted` and `getAsReactionCode` I wonder if we should kill this method off as it is kinda weird for `unicode` emoji. Name really only makes sense for CustomEmoji in conjunction with `getId`.

Reply:
I think its better to keep this for consistency, and `getFormatted` can be default implemented in `UnicodeEmoji` this way